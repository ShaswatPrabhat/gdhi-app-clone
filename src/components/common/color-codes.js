export default {
  getColorCodes() {
    // Update variables if changed.
    return [
      { score: "5", color: "#01C975", description: "Most Developed" },
      { score: "4", color: "#80E1CC", description: "" },
      { score: "3", color: "#FFE180", description: "" },
      { score: "2", color: "#FFCA82", description: "" },
      { score: "1", color: "#FCAB9C", description: "Least Developed" },
      { score: "NA", color: "#6C757D" },
    ];
  },
};
