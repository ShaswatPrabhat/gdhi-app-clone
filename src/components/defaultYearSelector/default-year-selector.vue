<script>
import Vue from "vue";
import axios from "axios";
import yearFilter from "./year-filter.vue";
import { EventBus } from "../common/event-bus";
import { EVENTS } from "../../constants";

export default Vue.extend({
  name: "defaultYearSelector",
  components: { yearFilter },
  data() {
    return {
      filteredYear: "",
      defaultYear: window.appProperties.getDefaultYear(),
      years: [],
    };
  },

  created() {
    this.fetchYears();
  },

  mounted() {
    EventBus.$on(EVENTS.YEAR_FILTERED, (filteredYear) => {
      this.filteredYear = filteredYear;
    });
  },

  methods: {
    fetchYears: function () {
      const self = this;
      axios.get("/api/bff/distinct_year").then(({ data }) => {
        self.years = data.years;
        this.defaultYear = data.defaultYear;
        window.appProperties.setDefaultYear({
          defaultYear: data.defaultYear,
        });
      });
    },

    setDefaultYear: function () {
      const self = this;

      let payload;
      if (self.filteredYear) {
        payload = self.filteredYear;
      } else {
        if (self.years.includes(this.defaultYear)) {
          payload = this.defaultYear;
        } else {
          payload = self.years[0];
        }
      }

      axios
        .post("/api/default_year/submit", payload, {
          headers: { "Content-Type": "text/plain" },
        })
        .then(() => {
          window.appProperties.setDefaultYear({
            defaultYear: this.filteredYear,
          });
        });
    },
  },
});
</script>

<template>
  <div>
    <div class="year-indicator">
      <div class="header-bold">
        Select year for which date is to be displayed on the Homepage
      </div>

      <yearFilter :selectedYear="defaultYear" :years="years" />
    </div>
    <button
      class="btn btn-primary"
      @click="setDefaultYear"
      style="margin-left: 1vw"
    >
      SUBMIT
    </button>
  </div>
</template>
