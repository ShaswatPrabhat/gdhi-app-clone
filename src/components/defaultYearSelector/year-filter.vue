<template>
  <select
    class="year-indicator-select"
    @change="filter"
    name="test_select3"
    :value="defaultYearValue"
  >
    <option v-for="(year, index) in years" :key="index" :value="year">
      {{ year }}
    </option>
  </select>
</template>

<script>
import { EventBus } from "../common/event-bus";
import { EVENTS } from "../../constants";

export default {
  name: "yearFilter",
  props: {
    years: Array,
    defaultYearValue: String,
  },

  methods: {
    filter: (event) => {
      EventBus.$emit(EVENTS.YEAR_FILTERED, event.target.value);
    },
  },
};
</script>
