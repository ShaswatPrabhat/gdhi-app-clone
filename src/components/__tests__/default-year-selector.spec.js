import { shallowMount } from "@vue/test-utils";
import { describe, it, expect, beforeEach, vi, afterEach } from "vitest";
import axios from "axios";
import defaultYearSelector from "../defaultYearSelector/default-year-selector.vue";
import { i18n } from "../../plugins/i18n";
import flushPromises from "flush-promises";
import { EventBus } from "../common/event-bus";

const axiosGetSpy = vi.spyOn(axios, "get");
const axiosPostSpy = vi.spyOn(axios, "post");
const eventBusOnSpy = vi.spyOn(EventBus, "$on");
const setDefaultYearSpy = vi.fn();
const getDefaultYearSpy = vi.fn();

window.appProperties = {
  setDefaultYear: setDefaultYearSpy,
  getDefaultYear: getDefaultYearSpy,
};
describe("year-filter", () => {
  let wrapper;

  beforeEach(async () => {
    axiosGetSpy.mockResolvedValueOnce({
      data: { years: ["Version 1", "2022", "2023"], defaultYear: "2022" },
    });
    wrapper = shallowMount(defaultYearSelector, { i18n });
    await flushPromises();
  });

  afterEach(() => {
    axiosGetSpy.mockReset();
    eventBusOnSpy.mockReset();
  });

  it("should render default Year Selector", async () => {
    expect(wrapper.vm.$el).toMatchSnapshot();
  });

  it("should call fetch years on mount and populate windows object", async () => {
    expect(axiosGetSpy.mock.calls[0][0]).to.equal("/api/bff/distinct_year");
    expect(setDefaultYearSpy.mock.calls[0][0].defaultYear).to.equal("2022");
  });

  it("register a listener on EventBus", async () => {
    expect(eventBusOnSpy.mock.calls[0][0]).to.equal("year:filtered");
  });

  it.skip("register a listener on EventBus which sets default year", async () => {
    eventBusOnSpy.mock.calls[0][1]("a_default_year");
    expect(wrapper.vm.defaultYear).to.equal("a_default_year");
  });

  it("trigger post API for default Year Change and populate windows object", async () => {
    axiosPostSpy.mockResolvedValue({});
    const submitButton = wrapper.find(".btn");

    submitButton.trigger("click");
    await flushPromises();
    expect(axiosPostSpy.mock.calls[0][0]).to.equal("/api/default_year/submit");
    expect(axiosPostSpy.mock.calls[0][1]).to.equal("2022");

    expect(setDefaultYearSpy.mock.calls[0][0].defaultYear).to.equal("2022");
  });
});
