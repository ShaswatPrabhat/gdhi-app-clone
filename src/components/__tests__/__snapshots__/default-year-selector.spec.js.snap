// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`year-filter > should render default Year Selector 1`] = `
<div>
  <div
    class="year-indicator"
  >
    <div
      class="header-bold"
    >
       Select year for which date is to be displayed on the Homepage 
    </div>
    <yearfilter-stub
      selectedyear="2022"
      years="Version 1,2022,2023"
    />
  </div>
  <button
    class="btn btn-primary"
    style="margin-left: 1vw;"
  >
     SUBMIT 
  </button>
</div>
`;
